[receiver/gitproviderreceiver] initial commit of git provider receiver skeleton

adrielp opened the pull request:
**Description:**
Add initial commit of receiver component including the github sub scraper. This receiver is in development and what is included here is only a skeleton.
**Link to tracking Issue:**  
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22028

**Testing:**

- basic tests were added but as this is in development, it's only a skeleton.

**Documentation:**

- basic set of documentation was added as part of the skeleton

Commit:
feat: initial commit of gitmetricsreceiver skeleton in development #22028

atoulme commented on receiver/gitmetricsreceiver/config.go:
```diff
@@ -0,0 +1,80 @@
+	"go.opentelemetry.io/collector/confmap"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitmetricsreceiver/internal"
+    "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitmetricsreceiver/internal/metadata"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
```
please run make lint in the folder.

atoulme commented:
Please add entry under `.github/CODEOWNERS`. Please run `make gendependabot` and `make generate-gh-issue-templates` from the root of the repository.

adrielp commented:
Just a heads up. I'm going to rebase & resolve the conflicts when I get back from working on a client site tomorrow/thursday. I also have just a couple lint issues remaining to fix.  Appreciate the suggestions & patience on this!

Commit:
fix: fix linting issues and remove unused vars from previous metrics

adrielp commented:
@astencel-sumo do i need to manually add you for review? I thought it would've added you since you're the sponsor. 

I've run through the lints, code owners, issue template, and dependabot commands. I've fixed some tests and things as well. 

This is still the skeleton, but it's a skeleton taken from the initial prototype we had working as a github receiver. Def more iterations, but I think it's good to review now. 

I'm hoping this isn't too much for a skeleton PR though.

adrielp commented on .github/CODEOWNERS:
Looks like there are several people here that don't have access to the repo as code owners so this is throwing errors.

adrielp commented:
Rebased to fix merge conflict in go.mod

astencel-sumo commented on receiver/gitmetricsreceiver/README.md:
```diff
@@ -0,0 +1,34 @@
+[development]: https://github.com/open-telemetry/opentelemetry-collector#development
+<!-- end autogenerated section --> 
+
+The Git Metrics receiver generates metrics from Git about Git repositories and 
```
```suggestion
The Git Metrics receiver generates metrics about Git repositories and 
```

The receiver generates metrics not from Git itself, but from various vendor endpoints, as you correctly describe in the end of this sentence. 🙂

astencel-sumo commented on receiver/gitmetricsreceiver/README.md:
```diff
@@ -0,0 +1,34 @@
+<!-- end autogenerated section --> 
+
+The Git Metrics receiver generates metrics from Git about Git repositories and 
+organizations from various vendor endpoints. It is intended to be used
```
Why would it only be "intended to be used when the collector is deployed as an agent"?

adrielp replied:
This is a poor choice of wording on my part. Good catch. The sentence as a whole doesn't need to exist at all. Will remove.

adrielp replied:
If we switch to `gitprovider` I'm thinking of switching this whole block to something very simple like:

> The Git Provider receiver scrapes data from git vendors.

Thoughts?

astencel-sumo commented on receiver/gitmetricsreceiver/README.md:
```diff
@@ -0,0 +1,34 @@
+
+## Getting Started
+
+The collection interval is defaulted to 30 seconds across the board. 
```
This sounds simpler to me, what do you think?

```suggestion
The collection interval is common to all scrapers and is set to 30 seconds by default.
```

astencel-sumo commented on receiver/gitmetricsreceiver/README.md:
````diff
@@ -0,0 +1,34 @@
+The collection interval is defaulted to 30 seconds across the board. 
+
+```yaml
+gitmetrics:
````
I'm still not sure if we got the receiver's name right - `gitmetrics`. I have a couple thoughts:

1. I'd rather get rid of the `metrics` part in the receiver name. It feels redundant to me.
2. I wonder if we should emphasize that this receiver's scope is "repository" metrics, and not CI/CD metrics that many Git vendors offer. Or perhaps we shouldn't specify the scope of the receiver in its name, as we may want to change it in the future? :thinking:

Brainstorming some names:
- `git` - someone might think it's metrics from the `git` binary? :thinking: 
- `gitprovider` - explicitly saying it's not metrics from `git` binary and only about Git itself, but from Git providers, whatever they provide
- `gitrepo` - narrows down the receiver's scope to only "Git repository" metrics
- `githealth` - narrows down the receiver's scope to only "Git repository health" metrics

adrielp replied:
This is a good callout. 

> Or perhaps we shouldn't specify the scope of the receiver in its name, as we may want to change it in the future?

I lean towards this being the case. This current set is just initial, and thinking towards the future, industry definitions like DORA, and the OTEP, I do think that it'd be better to look at this as a `gitprovider` receiver instead of `gitmetrics`.

`gitprovider` has my vote 👍🏼

astencel-sumo commented on receiver/gitmetricsreceiver/README.md:
```diff
@@ -0,0 +1,34 @@
+The available scrapers are:
+| Scraper  | Description             |
+|----------|-------------------------|
+| [github] | Git Metrics from GitHub |
```
We assume everyone knows what GitHub is, but let's still put the link there :wink:

```suggestion
| [github] | Git Metrics from [GitHub](https://github.com/) |
```

astencel-sumo commented on receiver/gitmetricsreceiver/config.go:
```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package gitmetricsreceiver // import "github.com/liatrio/otel-liatrio-contrib/receiver/githubreceiver"
```
```suggestion
package gitmetricsreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/githubreceiver"
```

astencel-sumo commented on receiver/gitmetricsreceiver/config.go:
```diff
@@ -0,0 +1,82 @@
+// Validate the configuration passed through the OTEL config.yaml
+func (cfg *Config) Validate() error {
+	if len(cfg.Scrapers) == 0 {
+		return errors.New("must specify at least one scraper when using git metrics receiver")
```
```suggestion
		return errors.New("must specify at least one scraper")
```

I know the `hostmetrics` receiver has similar message, but it's redundant:

```console
Error: invalid configuration: receivers::hostmetrics: must specify at least one scraper when using hostmetrics receiver
```

adrielp replied:
Yea, definitely leveraged the host metrics receiver when converting this to allow for multiple scrapers 😃

Commit:
docs: commit suggestion in receiver/gitmetricsreceiver/README.md

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

Commit:
docs: commit suggestion in readme about metric generation

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

Commit:
docs: commit suggestion to update readme to include github link

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

Commit:
fix: commit suggestion to update import comment with correct otel project

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

Commit:
chore: commit error message suggestion to remove redundancy

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

astencel-sumo commented:
You need to add a Changelog entry: [https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.80.0/CONTRIBUTING.md#adding-new-components).

Also the receiver needs to have its own `go.mod` file, otherwise it cannot be built into a distro even for testing.

Commit:
feat: rename gitmetrics to git provider based on feedback

Commit:
chore: add changelog entry

Commit:
chore: update codeowners file with new receiver name and update go mod

Commit:
chore: run generate template to update to new receiver name

astencel-sumo commented on receiver/gitproviderreceiver/factory.go:
```diff
@@ -0,0 +1,143 @@
+// This file implements a factory for the git provider receiver
+
+const (
+	typeStr         = "gitprovider"
```
You should not need this const `typeStr`, you should use `metadata.Type` instead.

astencel-sumo commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -0,0 +1,67 @@
+type: gitproviderreceiver
```
```suggestion
type: gitprovider
```

astencel-sumo commented on receiver/gitproviderreceiver/factory.go:
```diff
@@ -0,0 +1,143 @@
+	typeStr         = "gitprovider"
+	defaultInterval = 30 * time.Second
+	defaultTimeout  = 15 * time.Second
+	stability       = component.StabilityLevelDevelopment
```
You should not need this const `stability`, you should use `metadata.MetricsStability` instead.

astencel-sumo commented:
I was able to build a distro with this component by adding the following lines to my manifest:

```yaml
dist:
  name: otelcol-custom
  otelcol_version: 0.80.0
  output_path: ./build
  version: 0.80.0
exporters:
  - gomod: go.opentelemetry.io/collector/exporter/loggingexporter v0.80.0
  - gomod: go.opentelemetry.io/collector/exporter/otlpexporter v0.80.0
extensions:
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension v0.80.0
receivers:
  - gomod: go.opentelemetry.io/collector/receiver/otlpreceiver v0.80.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitproviderreceiver v0.80.0
replaces:
  - github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitproviderreceiver => ../../../opentelemetry-collector-contrib/receiver/gitproviderreceiver
```

and the following configuration:

```yaml
exporters:
  logging:
    verbosity: detailed

receivers:
  gitprovider:
    scrapers:
      githubscraper:

service:
  pipelines:
    metrics:
      exporters:
      - logging
      receivers:
      - gitprovider
```

The result was the following:

```console
./build/otelcol-custom --config ./config.yaml 
2023-06-27T12:10:40.989+0200    info    service/telemetry.go:81 Setting up own telemetry...
2023-06-27T12:10:40.989+0200    info    service/telemetry.go:104        Serving Prometheus metrics      {"address": ":8888", "level": "Basic"}
2023-06-27T12:10:40.989+0200    info    exporter@v0.80.0/exporter.go:275        Development component. May change in the future.        {"kind": "exporter", "data_type": "metrics", "name": "logging"}
2023-06-27T12:10:40.989+0200    info    receiver@v0.80.0/receiver.go:296        Development component. May change in the future.        {"kind": "receiver", "name": "gitprovider", "data_type": "metrics"}
2023-06-27T12:10:41.012+0200    info    service/service.go:131  Starting otelcol-custom...       {"Version": "0.80.0", "NumCPU": 16}
2023-06-27T12:10:41.012+0200    info    extensions/extensions.go:30     Starting extensions...
2023-06-27T12:10:41.012+0200    info    githubscraper/github_scraper.go:62      Starting the scraper inside scraper.go  {"kind": "receiver", "name": "gitprovider", "data_type": "metrics"}
2023-06-27T12:10:41.012+0200    info    service/service.go:148  Everything is ready. Begin running and processing data.
2023-06-27T12:10:43.374+0200    error   githubscraper/github_scraper.go:269     Error getting repos%!(EXTRA zapcore.Field={error 26 0  GET https://api.github.com/user/repos: 401 Requires authentication []})  {"kind": "receiver", "name": "gitprovider", "data_type": "metrics"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitproviderreceiver/internal/scraper/githubscraper.(*githubScraper).scrape
        github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gitproviderreceiver@v0.80.0/internal/scraper/githubscraper/github_scraper.go:269
go.opentelemetry.io/collector/receiver/scraperhelper.ScrapeFunc.Scrape
        go.opentelemetry.io/collector/receiver@v0.80.0/scraperhelper/scraper.go:20
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
        go.opentelemetry.io/collector/receiver@v0.80.0/scraperhelper/scrapercontroller.go:210
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
        go.opentelemetry.io/collector/receiver@v0.80.0/scraperhelper/scrapercontroller.go:188
2023-06-27T12:10:43.375+0200    info    MetricsExporter {"kind": "exporter", "data_type": "metrics", "name": "logging", "resource metrics": 1, "metrics": 1, "data points": 1}
2023-06-27T12:10:43.375+0200    info    ResourceMetrics #0
Resource SchemaURL: https://opentelemetry.io/schemas/1.9.0
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope otelcol/gitproviderreceiver 0.80.0
Metric #0
Descriptor:
     -> Name: git.repository.count
     -> Description: Number of repositories that exist in an organization
     -> Unit: 1
     -> DataType: Gauge
NumberDataPoints #0
StartTimestamp: 2023-06-27 10:10:40.989420537 +0000 UTC
Timestamp: 2023-06-27 10:10:42.013105112 +0000 UTC
Value: 0
        {"kind": "exporter", "data_type": "metrics", "name": "logging"}
^C2023-06-27T12:10:44.374+0200  info    otelcol/collector.go:236        Received signal from OS {"signal": "interrupt"}
2023-06-27T12:10:44.374+0200    info    service/service.go:157  Starting shutdown...
2023-06-27T12:10:44.374+0200    info    extensions/extensions.go:44     Stopping extensions...
2023-06-27T12:10:44.374+0200    info    service/service.go:171  Shutdown complete.
```

@adrielp, I propose to do one of the following:

1. Make the `github` scraper actually work and add the docs  on how to configure it
2. Remove the `github` scraper's code that cannot be run/tested and just have the scraper log an INFO level log saying "scraper started" or similar.

I'd actually prefer the second approach, as it would make it easier to review the `github` scraper's functionality in a separate PR. But if you prefer the first approach, I can go with that.

In the first approach, the scraper doesn't need to generate all the declared metrics - perhaps just one metric or all the "low hanging fruit" metrics that can be easily retrieved. More can be added later.

Commit:
chore: remove receiver from type name in metadata

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

Commit:
chore: change githubscraper to github in TypeStr for readability

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

adrielp commented:
> I was able to build a distro with this component with the following manifest and these [docs](https://opentelemetry.io/docs/collector/custom-collector/):
> 
> manifest.yaml
> and the following configuration:
> 
> config.yaml
> The result was the following:
> 
> console output
> @adrielp, I propose to do one of the following:
> 
> 1. Make the `github` scraper actually work and add the docs  on how to configure it
> 2. Remove the `github` scraper's code that cannot be run/tested and just have the scraper log an INFO level log saying "scraper started" or similar.
> 
> I'd actually prefer the second approach, as it would make it easier to review the `github` scraper's functionality in a separate PR. But if you prefer the first approach, I can go with that.
> 
> In the first approach, the scraper doesn't need to generate all the declared metrics - perhaps just one metric or all the "low hanging fruit" metrics that can be easily retrieved. More can be added later.

Yea, the ORG & auth creds have to be configured. 

I agree with approach 2 as well. I was certainly worried this was too much to review as part of the skeleton, but struggled to find the right stopping point given all the code I already had.

Commit:
chore: update pull request based on the following suggestions

* finish changing name within code from `gitproviderreceiver` to
  `gitprovider`
* fix the tests that broke due to that change
* update typestr to be pulled from metadata
* update stability to be pulled from metadata

Commit:
chore: remove github scraper functionality for skeleton pr

Removing this functionality without tests makes this initial skeleton pr
easier to review. This functionality will be added in subsequent prs.

Commit:
chore: fix linting and fmtting issues after code removal

Commit:
chore: run make tidy after removing code and dependencies

adrielp commented:
Removed the bulk of the `githubscraper` code, made adjustments based on feedback. 

Went and built the custom collector locally to confirm info logging change. Attached is a screenshot 
![image](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/25961386/6eb2ee15-e11f-4c02-b4fa-01992130cbb0)

astencel-sumo commented:
Can we somehow get rid of those EasyCLA comments about a commit not being linked to the GitHub account? They seem to pop up with every new push.

* :x: The email address for the commit ([f4e5f08](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f4e5f08d8711004407f5142bf03e42d57f505d53)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

Commit:
chore: run make goporto to fix ci failure

astencel-sumo commented:
All right, looking good. Thanks for this!

The last thing I would ask is to remove all the metrics, attribute, resource attribute definitions from the `metadata.yaml` file, as we're not generating any at the moment. Let's add those back together with the code to actually generate those metrics, does it make sense?

Other than that, run `make gotidy`, merge latest changes from `main` and we should be good to go. :clap:

adrielp commented:
> All right, looking good. Thanks for this!
> 
> The last thing I would ask is to remove all the metrics, attribute, resource attribute definitions from the `metadata.yaml` file, as we're not generating any at the moment. Let's add those back together with the code to actually generate those metrics, does it make sense?
> 
> Other than that, run `make gotidy`, merge latest changes from `main` and we should be good to go. 👏

Sure thing, can do that. When I remove those, do you want me to re-run `mdatagen` or is it fine to leave the generated `internal/metadata` code with the metrics in place?

Commit:
chore: remove metrics from metadata and re-gen based on feedback for
skeleton

* Will be added back with future iterations, removing now for skeleton
  pr

Commit:
chore: run make gotidy for the checks

adrielp commented:
* removed info in metadata and regenerated via mdatagen
* ran all the commands just to confirm
* rebased off latest in main
* ran make gotidy

I think we should be good 🤞🏼 @astencel-sumo

astencel-sumo commented:
> * removed info in metadata and regenerated via mdatagen

What's interesting, removing the metrics definitions from `metadata.yaml` and running `make generate` apparently did not remove the pre-existing `documentation.md` and other generated files.

In the end, I think I'll let that go, because apparently removing the generated files would mean also making changes in `factory.go` and other files to remove some autogenerated types, and this PR has already had so many back-and-forth that you're probably very eager to merge it and continue working on the actual scraping. :slightly_smiling_face:

astencel-sumo commented:
One last thing @adrielp - you need to run `make gendependabot` for all checks to pass.

Commit:
chore: run make gendependabot for checks

adrielp commented:
> What's interesting, removing the metrics definitions from `metadata.yaml` and running `make generate` apparently did not remove the pre-existing `documentation.md` and other generated files.

Yea I thought that was interesting too. I would've expected there to be empty generated files.

> In the end, I think I'll let that go, because apparently removing the generated files would mean also making changes in `factory.go` and other files to remove some autogenerated types, and this PR has already had so many back-and-forth that you're probably very eager to merge it and continue working on the actual scraping. 🙂

Ok, no worries. I'm just happy to be here getting a contribution started. There's been a good amount of conversation and learning. 

Ran make gendependabot. I must've forgotten to do that after changing the receiver name.

andrzej-stencel reviewed (approved):
:rocket:

astencel-sumo commented:
Approved @adrielp, thanks for your patience! :pray:

I think you need to merge `main` to get the checks to pass or sth?

adrielp commented:
Thank you for your patience with this PR as well, I know it's been a ton and not the easiest PR to review @astencel-sumo .

The [check-collector-module-version](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5422815071/jobs/9953136948?pr=23225) failed workflow is fixed locally. This was done by:
* rebasing again from latest
* re-running `go mod tidy` on the `gitproviderreceiver` to ensure up-to-date modules
* running `./.github/workflows/scripts/check-collector-module-version.sh` locally to verify 

The default [checks](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5422815071/jobs/9953136215?pr=23225) on the other hand, isn't fixed via that process. It looks to be running `make multimod-verify` which, if I'm reading this right on a quick skim, is wanting the `gitproviderreceiver` to be listed in the `versions.yaml` file. My understanding was that this would be part of actually putting the receiver in the bistro, which it's not. I did run `make multimod-sync` locally to see if that would fix it, and noticed that it actually references the `versions.yaml` file in a locally directory. Specifically it looks up one directory from the root of this repo, into the `opentelemetry-collector/versions.yaml` repo instead as a reference, requiring both repo's to be cloned down. Haven't dug farther into this yet, but will keep looking. 

Just a heads up, I will be slow to respond this week as I'm out of town. Will try to take a look in the evenings, but might be able to take an in-depth look till Monday.

Commit:
fix(ci): add gitproviderreceiver to versions.yaml

Commit:
Merge branch 'main' into gitmetricsreceiver-skeleton

adrielp commented:
@astencel-sumo  I think i fixed both of them. I figured out a step I missed (updating versions.yaml) and was able to get the collector module version check & then mod validate to run successfully, also resolved a merge conflict. Checks should pass this time.

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
fix: fix go module versions and regen dependabot

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
chore: run make generate again and merge latest from main

adrielp commented:
Try #12093102983. I might be missing something, but would be nice to add some of these make commands into the docs & add a local make command to run the entire test locally.

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
chore: update modules and merge in latest from main after fixes

Commit:
Merge branch 'main' into gitmetricsreceiver-skeleton

Commit:
Merge branch 'main' into gitmetricsreceiver-skeleton

Commit:
Merge branch 'main' into gitmetricsreceiver-skeleton

jpkrohling commented:
CI started again.

adrielp commented:
thanks @jpkrohling! would you mind merging?

jpkrohling commented on .github/dependabot.yml:
```diff
@@ -1097,8 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
```
Why is this being removed?

adrielp replied:
This was auto generated after merging the latest from main (a few days ago) and re-running all the generate commands. Looking at main, I'm not sure why the generate command is removing it because I still see sshcheck in main. I'm happy to merge again & re-run to ensure it gets added back.

adrielp replied:
Looking at this more, it looks to have been removed as a sideeffect of having placed a limit on [make gendependabot](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/Makefile#L144-L169)

As I understand, it's going to iterate through all the mod dirs, and it's going to stop appending to the generate file once you've got 220 entries. Looks like the addition of this component is bumping the other component out of place. 

The [build and test workflow](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/.github/workflows/build-and-test.yml#L209-L212) is going to run `make gendependabot` every time and error out if the diff is different. That's why the checks pass with the removal of that component, but don't pass if I haven't run the command & committed the change. 

I couldn't find any specific reference in GitHub's docs about a limit to 220 entries though. Happy to open an issue on this one & find out some more information.

jpkrohling replied:
> Happy to open an issue on this one & find out some more information.

That'd be great, thanks!

adrielp replied:
Issue has been created: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24363

jpkrohling commented on .github/CODEOWNERS:
```diff
@@ -190,6 +190,7 @@ receiver/filereceiver/                                   @open-telemetry/collect
 receiver/filestatsreceiver/                              @open-telemetry/collector-contrib-approvers @atoulme
 receiver/flinkmetricsreceiver/                           @open-telemetry/collector-contrib-approvers @jonathanwamsley @djaglowski
 receiver/fluentforwardreceiver/                          @open-telemetry/collector-contrib-approvers @dmitryax
+receiver/gitproviderreceiver/                            @open-telemetry/collector-contrib-approvers @adrielp @astencel-sumo
```
@atoulme, are we enforcing that new code owners have to be members of OTel?

adrielp replied:
As soon as this pr is through, I'll be opening up a membership request. I'm currently looking for one more sponsor 😉 <cough><cough>

jpkrohling commented on receiver/gitproviderreceiver/README.md:
```diff
@@ -0,0 +1,32 @@
+[development]: https://github.com/open-telemetry/opentelemetry-collector#development
+<!-- end autogenerated section --> 
+
+The Git Provider receiver scrapes data from Git vendors.
```
I'm having some trouble understanding what this means just by looking at the readme. Can you provide a few examples here? Like: 

"When GitHub is used, the following metrics are expected to be created as part of this receiver:"

It's OK to include only a couple of examples, pointing people to the documentation.md file for complete reference.

adrielp replied:
Updated this readme with additional information & regenerated as we've created our own upstream distro of the collector.

jpkrohling commented on receiver/gitproviderreceiver/config.go:
```diff
@@ -0,0 +1,82 @@
+	for key := range scrapersSection.ToStringMap() {
+		factory, ok := getScraperFactory(key)
+		if !ok {
+			return fmt.Errorf("invalid scraper key: %s", key)
```
```suggestion
			return fmt.Errorf("invalid scraper key: %q", key)
```

jpkrohling commented on receiver/gitproviderreceiver/config.go:
```diff
@@ -0,0 +1,82 @@
+		}
+
+		collectorCfg := factory.CreateDefaultConfig()
+		collectorViperSection, err := scrapersSection.Sub(key)
```
Viper?

adrielp replied:
I'm certain that's from copy pasta'ing the format from the host metrics receiver when converting this into a receiver with multiple scrapers vs one specific vendor scraper.

adrielp replied:
Removed `Viper` from the name.

jpkrohling commented on receiver/gitproviderreceiver/README.md:
````diff
@@ -0,0 +1,32 @@
+The collection interval is common to all scrapers and is set to 30 seconds by default.
+
+```yaml
+gitprovider:
````
Please include the typical options people would use to get started with this, even if it's not implemented yet (it's marked as "development" after all). If GitHub is what we provide today, perhaps give a full example of GitHub, including how to set the token (might involve other extensions for that?)

jpkrohling commented on receiver/gitproviderreceiver/factory.go:
```diff
@@ -0,0 +1,136 @@
+			return nil, fmt.Errorf("failed to create scraper %q: %w", key, err)
+		}
+
+		if ok {
```
Perhaps invert the logic? If not ok, return the error, then end the for loop with the success case (the line below)?

adrielp replied:
Inverted logic

jpkrohling commented on receiver/gitproviderreceiver/factory.go:
```diff
@@ -0,0 +1,136 @@
+	key string,
+	cfg internal.Config,
+	factories map[string]internal.ScraperFactory,
+) (scraper scraperhelper.Scraper, ok bool, err error) {
```
Why do you need both a "ok" flag and an error? If the error is not nil, isn't it automatically not ok? Are there scenarios where OK is true and error is not nil, or that OK is false and the error is nil?

adrielp replied:
Yes, there are scenarios. This was mimic'd from the hostmetricsreceiver as well during the transition to multiple scrapers instead of just one. The logic here is that the calling function will check whether it's an error from creating a scraper or if the scraper just doesn't exist, and return an error message accordingly. 

The scraper could exist, but fail to create (that would be err) or it could not be ok, in that the scraper isn't a actual scraper. 

We could refactor this such that the error message is just returned from the `createGitProviderScraper` directly up, thereby not needing to do the error & ok checking at the calling function. From a troubleshooting perspective, I don't think it's important to log both possibilities though.

jpkrohling replied:
> The logic here is that the calling function will check whether it's an error from creating a scraper or if the scraper just doesn't exist, and return an error message accordingly

There could be two different errors, and the caller here would just handle the errors accordingly?

adrielp replied:
updated the logic to remove `ok` entirely

jpkrohling commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func (ghs *githubScraper) start(_ context.Context, host component.Host) (err error) {
+	ghs.logger.Sugar().Info("Starting the scraper inside scraper.go")
```
Do you have the scraper name already as part of the logger context? If not, include the scraper name. And the file name seems outdated. From a UX perspective, I would write something like:

```suggestion
	ghs.logger.Info("Starting the GitHub scraper", zap.String("host", ghs.cfg.Endpoint))
```

adrielp replied:
This was something I added when I was first learning how to develop a receiver. I agree, it's message can either be removed or improved in general to be more informative to the user reading the logs.

adrielp replied:
I've updated the message to just say `starting the GitHub scraper`. Right now the library making the calls to GitHub presumes publicly available github instead of hosted, so Endpoint isn't something provided by default. Will adjust this message accordingly in future iterations for clarity as I additionally create test cases for privately hosted enterprise github via custom endpoints.

jpkrohling commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	now := pcommon.NewTimestampFromTime(time.Now())
+	ghs.logger.Sugar().Debugf("current time: %v", now)
```
```suggestion
	ghs.logger.Sugar().Debug("current time", zap.Time("now", now.AsTime()))
```

Same for the other similar log.

jpkrohling commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -0,0 +1,70 @@
+	currentDate := time.Now().Day()
+	ghs.logger.Sugar().Debugf("current date: %v", currentDate)
+
+	ghs.logger.Sugar().Info("github scraper has started")
```
Seems like a debug to me.

adrielp replied:
Yea absolutely, Same comment as from above.

jpkrohling reviewed (commented):
It would have been helpful to include what the next steps are, but the PR looks good other than a few details.

Commit:
chore: update error message with quoted string formatter based on suggestion

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
chore: address documentation feedback and update distro

Commit:
chore: update debug statement with suggested changes

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

Commit:
chore: update logging statements and code based on suggestions

jpkrohling commented on receiver/gitproviderreceiver/factory.go:
```diff
@@ -0,0 +1,135 @@
+			return nil, fmt.Errorf("failed to create scraper %q: %w", key, err)
+		}
+
+        if !ok {
```
The indentation here looks odd; it'll likely fail the linter.

adrielp replied:
ran fmt and lint locally

jpkrohling reviewed (commented):
LGTM, there's only a question of returning three values from a function, and the indentation. Otherwise, I would have started the CI already.

Commit:
chore: run fmt and lint

Commit:
Merge remote-tracking branch 'origin/main' into gitmetricsreceiver-skeleton

Commit:
chore: run go mod tidy on gitprovider receiver after merge

Commit:
chore: change scraper error handling and update test

Commit:
chore: refmt after updates

Commit:
Merge branch 'main' into gitmetricsreceiver-skeleton

Commit:
chore: fix linting issue and add liatrio distro to mdatagen

* there's no auto linking of liatrio distro right now, so make generate
  causes an invalid link to be placed, failing the linter

jpkrohling reviewed (commented):
LGTM. I confirmed with the other maintainers that we are indeed requiring code owners to be members of the organization. Once you are one, you can be included in the code owners. I'm fine being one sponsor for your membership, but you'll need another one as well.

adrielp commented:
> LGTM. I confirmed with the other maintainers that we are indeed requiring code owners to be members of the organization. Once you are one, you can be included in the code owners. I'm fine being one sponsor for your membership, but you'll need another one as well.

Ok. I'll go ahead and enter that now and post it. @astencel-sumo already volunteered to be one of my sponsors so I'll include y'all now. 

side note: looks like there were general API issues with uploading the code coverage report, unrelated to any changes. might be an outage or a fluke, not sure if that's one of those things that is flaky sometimes.

jpkrohling commented:
Once your membership request has been approved, let me know and I'll merge this.

Commit:
chore: merge in latest from main and run generation commands

Commit:
chore: fix metadata owners and broken markdown links to pass linting

Commit:
chore: merge in latest and resolve conflicts, regenerate
